src/osd/PrimaryLogPG.cc

```diff
@@ -12443,15 +12443,14 @@ bool PrimaryLogPG::start_recovery_ops(
 
   const auto &missing = pg_log.get_missing();
 
-  unsigned int num_missing = missing.num_missing();
   uint64_t num_unfound = get_num_unfound();
 
-  if (num_missing == 0) {
+  if (!missing.have_missing()) {
     info.last_complete = info.last_update;
   }
 
-  if (num_missing == num_unfound) {
-    // All of the missing objects we have are unfound.
+  if (!missing.have_missing() || // Primary does not have missing
+      all_missing_unfound()) { // or all of the missing objects are unfound.
```

Comment:
@xiexingguo Considering the scenario described in http://tracker.ceph.com/issues/38784, this will help to recover the missing object on the primary, before calling recover_replicas(), but wouldn't `all_missing_unfound()` (when we have 1 unfound object that is not accounted for in the missing) return false, when checking https://github.com/ceph/ceph/pull/27205/commits/d9497139a6f516ce015bb43a2e7f8958638cf8f8#diff-d43117703c33bebe41fc39224c8025adR1614?